vividus/src/main/java/org/vividus/transformer/InnerJoinTableTransformer.java

```diff
@@ -0,0 +1,87 @@
+ * limitations under the License.
+ */
+
+package org.vividus.transformer;
+
+import static org.apache.commons.lang3.Validate.isTrue;
+
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+import java.util.Set;
+import java.util.stream.Collectors;
+
+import org.jbehave.core.model.ExamplesTable;
+import org.jbehave.core.model.ExamplesTable.TableProperties;
+import org.jbehave.core.model.TableParsers;
+import org.vividus.util.ExamplesTableProcessor;
+
+public class InnerJoinTableTransformer extends AbstractTableLoadingTransformer
+{
+    public InnerJoinTableTransformer()
+    {
+        super(false);
+    }
+
+    @Override
+    public String transform(String tableAsString, TableParsers tableParsers, TableProperties properties)
+    {
+        List<ExamplesTable> tables = loadTables(tableAsString, properties);
+        isTrue(tables.size() == 2, "Please, specify only two ExamplesTable-s");
+        ExamplesTable leftTable = tables.get(0);
+        ExamplesTable rightTable = tables.get(1);
+        String leftTableJoinColumn = properties.getMandatoryNonBlankProperty("leftTableJoinColumn", String.class);
+        String rightTableJoinColumn = properties.getMandatoryNonBlankProperty("rightTableJoinColumn", String.class);
+        isTrue(leftTable.getHeaders().contains(leftTableJoinColumn),
+                "The left table doesn't contain the following column: %s", leftTableJoinColumn);
+        isTrue(rightTable.getHeaders().contains(rightTableJoinColumn),
+                "The right table doesn't contain the following column: %s", rightTableJoinColumn);
+        Set<String> repeatingKeys = leftTable.getHeaders().stream()
+                .filter(e -> !e.equals(leftTableJoinColumn) && rightTable.getHeaders().contains(e))
+                .collect(Collectors.toSet());
+        isTrue(repeatingKeys.isEmpty(), "Tables must contain different columns (except joint column),"
+                + " but found the same columns: %s", repeatingKeys);
+
+        ExamplesTable resultTable = innerJoin(leftTable, rightTable, leftTableJoinColumn, rightTableJoinColumn);
+        return resultTable.isEmpty() ? getEmptyTableWithHeaders(tables, properties) : resultTable.asString();
+    }
+
+    private static ExamplesTable innerJoin(ExamplesTable leftTable, ExamplesTable rightTable,
+            String leftTableJoinColumn, String rightTableJoinColumn)
+    {
+        List<Map<String, String>> leftRows = leftTable.getRows();
+        List<Map<String, String>> rightRows = rightTable.getRows();
+        Map<String, List<Map<String, String>>> rightByColumn = rightRows.stream()
+                .collect(Collectors.groupingBy(m -> m.get(rightTableJoinColumn)));
+        List<Map<String, String>> examplesTableRows = leftRows.stream()
+                .filter(row -> rightByColumn.containsKey(row.get(leftTableJoinColumn)))
+                .flatMap(leftRow -> rightByColumn.get(leftRow.get(leftTableJoinColumn)).stream()
```

Comment:
it should be reworked to optimize the code: `Map#containsKey` and `Map#get` are used together and do search through the map, but it should be enough to perform only 1 search in the map

Reply:
changed
other version:
```
        List<Map<String, String>> examplesTableRows = leftRows.stream()
                .flatMap(leftRow ->
                {
                    List<Map<String, String>> value = rightByColumn.get(leftRow.get(leftTableJoinColumn));
                    return value == null ? null : value.stream()
                                    .map(rightRow ->
                                    {
                                        Map<String, String> jointRow = new HashMap<>();
                                        jointRow.putAll(leftRow);
                                        jointRow.putAll(rightRow);
                                        return jointRow;
                                    });
                }).collect(Collectors.toList());
```